crates/iota-network/src/discovery/mod.rs

```diff
@@ -451,6 +456,67 @@ impl DiscoveryEventLoop {
     }
 }
 
+/// Verifies that a peer actually controls the addresses they claim by
+/// attempting to establish a connection. This prevents address spoofing
+/// attacks. Tries all addresses individually and returns the list of
+/// addresses that were successfully verified.
+/// This should only be used for not-yet-connected peers.
+async fn verify_address_ownership(
+    network: &Network,
+    peer_info: &SignedNodeInfo,
+    config: &DiscoveryConfig,
+) -> Vec<Multiaddr> {
+    // Try each address individually and collect the ones that work
+    let verification_futures: Vec<_> = peer_info
+        .addresses
+        .iter()
+        .map(|address| {
+            let network = network.clone();
+            let peer_id = peer_info.peer_id;
+            let address = address.clone();
+            Box::pin(async move {
+                if let Ok(anemo_address) = address.to_anemo_address() {
+                    // Try to connect to the claimed address with the claimed peer_id
+                    match tokio::time::timeout(
+                        config.address_verification_timeout(),
+                        network.connect_with_peer_id(anemo_address, peer_id),
```

Comment:
We really need to be careful here. What happens if we are already connected (or almost connected, since the process of connecting in anemo network is asynchronous) to this peer (a peer with `peer_id`)? There will be `simultaneous_dial_tie_breaking`. So only one connection will survive.